src/libraries/System.Private.CoreLib/src/System/DateOnly.cs

```diff
@@ -0,0 +1,902 @@
+        private const int MaxDayNumber = 3652058;
+
+        private static int DayNumberFromDateTime(DateTime dt) => (int)(dt.Ticks / TimeSpan.TicksPerDay);
+
+        private DateTime GetEquivalentDateTime() => new DateTime(_dayNumber * TimeSpan.TicksPerDay);
+
+        private DateOnly(int dayNumber)
+        {
+            Debug.Assert((uint)dayNumber <= 3652058);
+            _dayNumber = dayNumber;
+        }
+
+        /// <summary>
+        /// Return the instance of the DateOnly structure representing the minimal possible date can be created.
+        /// </summary>
+        public static DateOnly MinValue => new DateOnly(MinDayNumber);
+
+        /// <summary>
+        /// Return the instance of the DateOnly structure representing the maximal possible date can be created.
+        /// </summary>
+        public static DateOnly MaxValue => new DateOnly(MaxDayNumber);
+
+        /// <summary>
+        /// Initializes a new instance of the DateOnly structure to the specified year, month, and day.
+        /// </summary>
+        /// <param name="year">The year (1 through 9999).</param>
+        /// <param name="month">The month (1 through 12).</param>
+        /// <param name="day">The day (1 through the number of days in month).</param>
+        public DateOnly(int year, int month, int day) => _dayNumber = DayNumberFromDateTime(new DateTime(year, month, day));
+
+        /// <summary>
+        /// Initializes a new instance of the DateOnly structure to the specified year, month, and day for the specified calendar.
+        /// </summary>
+        /// <param name="year">The year (1 through the number of years in calendar).</param>
+        /// <param name="month">The month (1 through the number of months in calendar).</param>
+        /// <param name="day">The day (1 through the number of days in month).<paramref name="month"/>.</param>
+        /// <param name="calendar">The calendar that is used to interpret year, month, and day.<paramref name="month"/>.</param>
+        public DateOnly(int year, int month, int day, Calendar calendar) => _dayNumber = DayNumberFromDateTime(new DateTime(year, month, day, calendar));
+
+        /// <summary>
+        /// Initializes a new instance of the DateOnly structure to the specified number of days.
+        /// </summary>
+        /// <param name="dayNumber">The number of days since January 1, 0001 in the Gregorian calendar.</param>
+        public static DateOnly FromDayNumber(int dayNumber)
+        {
+            if ((uint)dayNumber > MaxDayNumber)
+            {
+                ThrowHelper.ThrowArgumentOutOfRange_DayNumber(dayNumber);
+            }
+
+            return new DateOnly(dayNumber);
+        }
+
+        /// <summary>
+        /// Gets the year component of the date represented by this instance.
+        /// </summary>
+        public int Year => GetEquivalentDateTime().Year;
```

Comment:
Unless you mean caching the result, then probably.  The other option would be moving to separated Y/M/D values.

Reply:
Yes we need to have the struct size small as possible this will have a huge perf benefit passing the struct around. this is the same idea we do with DateTime.